Reduce references to mpmath internal code

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

https://github.com/mpmath/mpmath/issues/704#issuecomment-1732273893

#### Brief description of what is fixed or changed

Reduce references to mpmath internals throughout the codebase. Many places use some particular functions such as prec_to_dps so we import those in one place in a new `sympy.external.mpmath` module. If needed those functions could be copied from mpmath to avoid importing from libmp.

Also Poly.nroots is changed to use a local context rather than messing around with the global precision. Other places in the codebase that use mpmath directly even if not using the libmp stuff should often be changed to use local contexts so that is maybe something that could be standardised a bit better with perhaps helpers in the mpmath module.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
    * Various parts of the codebase have been changed to avoid manipulating mpmath's global precision. This should improve thread-safety for SymPy's core functionality like evalf.
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
maint: centralise references to mpmath.libmp

Many modules in SymPy were directly importing from mpmath's internal
code. This reduces some references to mpmath and centralises others so
that there are at least only a few places directly importing from
mpmath.libmp.

oscarbenjamin commented:
This is ready for review.

asmeurer commented on sympy/polys/polytools.py:
```diff
@@ -3711,42 +3722,31 @@ def nroots(f, n=15, maxsteps=50, cleanup=True):
-                    for coeff in f.all_coeffs()]
-            with mpmath.workdps(n):
+            coeffs = [coeff.evalf(n=n).as_real_imag() for coeff in f.all_coeffs()]
+            with ctx.workdps(n):
```
How does this differ from the mpmath.workdps?

oscarbenjamin replied:
The `ctx.workdps` context method temporarily changes the precision of the context `ctx`. The `mpmath.workdps` function is just the workdps method for the global context:
https://github.com/mpmath/mpmath/blob/8a2bed58a5dfa8cc90b06bba373ecaade7129422/mpmath/__init__.py#L39

By using `ctx.workdps` we are not manipulating the global context which is e.g. not thread-safe. The global context should be thought of as belonging to the end user rather than being something that a library like sympy can manipulate.

This here is not thread-safe either because I have added a context that is global in this module but at least it does not clash with anything else that might be using mpmath. Unfortunately just creating a context as a temporary local variable is quite slow:
```python
In [8]: %timeit ctx = MPContext()
562 μs ± 17.8 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [9]: %timeit ctx.cos(1)
9.61 μs ± 36.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
This means that it would be inefficient to create a context every time we want to call only one method. I am still not sure what the best solution for this is but we likely need a cache of contexts that we can pull from on a thread-safe way.

There are a bunch of other places around the sympy codebase that manipulate the mpmath global context and that are not changed in this PR but they should all be changed to look like the code here.

asmeurer replied:
OK, I assumed this had something to do with thread safety which was why I was confused why it was still globally cached. 

This really feels like something that should be improved in mpmath itself.

oscarbenjamin replied:
> I have a followup prepared to build on this PR that adds a `local_workprec` context manager so you can do:

I've just pushed that here now.

oscarbenjamin committed:
fix: remove all imports from mpmath.libmp submodules

oscarbenjamin committed:
fix: centralise all mpmath.libmp imports

haru-44 commented on sympy/external/ntheory.py:
```diff
@@ -79,17 +84,22 @@ def remove(x, f):
+    return int(_ifac(int(x)))
+
+
+def fibonacci(n):
```
If we follow the approach used for other functions in `sympy.external.ntheory`, I think it would be better to add `sympy.external.gmpy.fib` and use that instead.

haru-44 commented on sympy/polys/ring_series.py:
```diff
@@ -47,14 +47,14 @@
 from sympy.polys.monomials import (monomial_min, monomial_mul, monomial_div,
                                    monomial_ldiv)
-from mpmath.libmp.libintmath import ifac
+from sympy.external.ntheory import factorial as ifac
```
Is there any particular reason why `sympy.external.gmpy` isn’t being used?

oscarbenjamin committed:
Centralise all mpmath imports and use local contexts.

Adds thread-safe local_workprec and local_workdps context managers for
efficiently creating an mpmath context with a given precision. These are
used in place of workprec and workdps in most places around the codebase
to avoid manipulating mpmath's global precision.

Also centralise all imports from mpmath to sympy.external.mpmath so we
have everything in one place.

oscarbenjamin committed:
fix(matrices): fix precision in eigenvects

oscarbenjamin committed:
fix(mpmath): Add the interval context for IntervalPrinter

oscarbenjamin committed:
fix(evalf): use the correct contexts in evalf_integral and evalf_sum

oscarbenjamin committed:
fix(mpmath): use local context methods for mpmath functions.

When using a local context it is important to write e.g.

  ctx.add(a, b)

rather than

  a + b

The latter form cannot be trusted to use the correct precision.

oscarbenjamin commented:
This is ready for review again.

I've pushed a bunch more changes:

- Centralise all use of mpmath to the external/mpmath module.
- Add local_workpec context manager to create a temporary context that does not use mpmath's global precision.
- Eliminate almost all places that depend on the global precision by using local contexts.

There are a few places remaining that still use mpmath's global context but I think it would be better to handle those in a followup.

Not using the global precision means that code that uses mpmath needs to do e.g.:
```
with local_workprec(prec) as mp:
    v = mp.fadd(x, y)
```
rather than
```
with workprec(prec):
    v = x + y
```
Using `x + y` means that we don't have good control over what context is used for the calculation and only works correctly if everything uses the global context. Even just `mpmath.mpf(num)` depends on the global context.

Some places use `lambdify` to create a function that uses mpmath's global context but we can use the local context as the module:
```
with local_workprec(prec) as mp:
    f = lambdify(x, expr, modules=mp)
    v = f(x)
```
This works with changing the precision like `mp.prec = 100`.

oscarbenjamin committed:
fix(polys): use a local mpmath context in nroots

This avoids messing with the precision of the global mpmath context.

oscarbenjamin committed:
Remove unused import

oscarbenjamin committed:
external: add dummy function for type checker

oscarbenjamin commented:
I split out the first part of this into gh-28764 which is now merged.

The second part is in gh-28765 now so I'll close this.
